### ISSUE TITLE
[RFC]: Unified Data Parallelism

### ISSUE THREAD

**Issue by evanchueng** [vllm-project/vllm-omni#5494]
### Motivation.

# RFC: Unified Data Parallelism for Cosmos3

## Motivation

Cosmos3 supports a wide range of inference scenarios: text-to-image (T2I), text-to-video (T2V), image-to-video (I2V), video generation with audio, and action-conditioned generation. These use cases impose diverse latency, memory, and hardware requirements, especially between the Nano and Super variants. To serve all efficiently, we need a parallelization strategy that is both resource-aware and scalable.

## DP vs. SP: Fundamental Performance Comparison

Data Parallelism (DP) and Sequence Parallelism (SP) represent fundamentally different approaches to distributed inference:

| Aspect | Data Parallelism (DP) | Sequence Parallelism (SP) |
|--------|----------------------|--------------------------|
| Parallelism dimension | Batch (different prompts) | Sequence (same prompt split across ranks) |
| Communication pattern | None (independent replicas) | All-to-All over activations |
| Communication volume | Zero inter-rank | O(batch × seqlen × hiddendim) |
| GEMM efficiency | Full sequence → high utilization | Partial sequence/head → low utilization |
| Load balance | Problematic (varying prompt lengths) | Perfect (equal splits of one sequence) |
| Memory per rank | Full model + full sequence | Full model + (1 / DP_size) of the sequence |

### Key Bottlenecks by Metric

| Metric | SP | DP | Winner |
|--------|----|----|--------|
| Inter-rank communication | O(seq_len) All-to-All | Zero | DP |
| GEMM utilization | Low (partial sequence) | High (full sequence) | DP |
| Load balance | Perfect | Problematic (bubble) | SP |
| Throughput scaling | Sub-linear | Near-linear (bubble-limited) | DP (if bubble solved) |

**Conclusion:** DP offers superior communication and compute characteristics, but it suffers from memory pressure and the bubble problem caused by uneven sequence lengths.

### Challenge 1: Device HBM Bottleneck

Each DP rank holds the full model. Cosmos3-Super (64B parameters, FP16 ~129 GB) exceeds the 80 GB HBM of an H100 GPU.

### Challenge 2: CPU Host Memory Bottleneck

Traditional DP stores a full model copy per rank in host memory:

| DP Size | Host Memory Required | Feasibility |
|---------|---------------------|-------------|
| 8× GPUs | 8 × 129 GB = 1,032 GB | ❌ Impractical |

### Challenge 3: Activation Memory Pressure

For long sequences (e.g., 648K tokens), per-layer activation memory exceeds 40 GB (attention QKV + LN/RoPE + MoE intermediates).

### Challenge 4: The Bubble Problem

With varying prompt lengths, short-sequence ranks finish early and idle:

| Rank | Prompt Type | Frames | Compute Time | Idle Time |
|------|------------|--------|-------------|-----------|
| GPU 0 | T2I | 1 | 100 ms | 140 ms |
| GPU 1 | T2V (short) | 81 | 180 ms | 60 ms |
| GPU 2 | T2V (medium) | 189 | 240 ms | 0 ms |
| GPU 3 | T2V (medium) | 189 | 240 ms | 0 ms |

Bubble rate = 20.8%

> Note: SP avoids this problem because all ranks process the same sequence split evenly.

## Current DP Implementation in vLLM-Omni

| Component | Status | Evidence |
|-----------|--------|----------|
| _DIT GroupCoordinator | ✅ Exists | `parallel_state.py` defines `_DIT` as "VAE/DiT specific data parallel group" |
| DP rank tracking | ✅ Exists | GPUARWorker tracks `data_parallel_rank_local` and `data_parallel_index` |
| Complete DP execution flow | ✅ Merged | [PR #4707](https://github.com/vllm-project/vllm-omni/pull/4707) |

How the current design handles bubbles:
- DP is modelled as "independent replicas" – idle time does not block others.
- Load balancing is delegated to upper-layer request routing.

## Proposed Solution: Unified DP Architecture

We propose four interdependent sub-RFCs to unlock the full potential of DP:

```
┌─────────────────────────────────────────────────────────────────────────────┐
│                    Unified DP Architecture                                 │
├─────────────────────────────────────────────────────────────────────────────┤
│                                                                             │
│  ┌─────────────────────────────────────────────────────────────────────┐   │
│  │         Sub-RFC 1: Distributed Layerwise Offload                   │   │
│  │  Addresses: HBM + Host Memory Bottlenecks                         │   │
│  │  • Shard weights → host memory: 1,032 GB → 129 GB (constant)     │   │
│  │  • Double-buffer → HBM: 129 GB → 3.2 GB (2 layers)               │   │
│  │  • mmap weight loading → 0 private RSS (shared page cache)        │   │
│  └─────────────────────────────────────────────────────────────────────┘   │
│                                      │                                      │
│                                      ▼                                      │
│  ┌─────────────────────────────────────────────────────────────────────┐   │
│  │         Sub-RFC 2: Data Parallelism for Video Generation          │   │
│  │  Addresses: DP Execution + Bubble Problem                        │   │
│  │  • DP multi-concurrency → N concurrent requests                   │   │
│  │  • Dynamic load balancing → bubble rate < 5%                      │   │
│  └─────────────────────────────────────────────────────────────────────┘   │
│                                      │                                      │
│                                      ▼                                      │
│  ┌─────────────────────────────────────────────────────────────────────┐   │
│  │         Sub-RFC 3: Iterative Activation Processing                 │   │
│  │  Addresses: Activation Memory Pressure                             │   │
│  │  • Per-head iteration → attention: 33.6 GB → 1.05 GB             │   │
│  │  • Per-chunk iteration → MoE: 39.3 GB → 0.27 GB                  │   │
│  └─────────────────────────────────────────────────────────────────────┘   │
│                                      │                                      │
│                                      ▼                                      │
│  ┌─────────────────────────────────────────────────────────────────────┐   │
│  │         Sub-RFC 4: Ref-Target KV Cache with Layer-Wise Caching    │   │
│  │  Addresses: KV Cache + Redundant UND Compute                      │   │
│  │  • Cache Ref K/V per layer at T0, reuse in subsequent steps       │   │
│  │  • Bypass reference self-attention after first step               │   │
│  │  • Reduce per-step cross-attention compute by ~50%                │   │
│  └─────────────────────────────────────────────────────────────────────┘   │
│                                                                             │
└─────────────────────────────────────────────────────────────────────────────┘
```

### Sub-RFC 1: Distributed Layerwise Offload with H2D + AllGather Overlap

| Item | Link | Status |
|------|------|--------|
| RFC | [Issue #5396](https://github.com/vllm-project/vllm-omni/issues/5396) | Open |
| PR | [PR #5397](https://github.com/vllm-project/vllm-omni/pull/5397) | **Waiting for merge** |

**Target:** HBM and host memory bottlenecks.

**Key mechanisms:**
- Shard model weights across DP ranks and store them in host memory via mmap views (aggregate host usage drops from 1,032 GB to a constant ~129 GB for 8 ranks)
- Use double-buffering to keep only 2 layers on device at any time, reducing per-rank HBM occupancy from 129 GB to ~3.2 GB
- Overlap asynchronous H2D transfers and AllGather with computation on dedicated streams
- mmap weight loading: weights loaded as mmap views pointing to shared OS page cache, eliminating O(dp_size × model_size) RSS during model creation
- DP multi-concurrency: N concurrent requests with near-linear throughput scaling (AllGather only gathers weight shards, request-independent)

**Zero model-specific code changes:** The offloader handles meta conversion (`to_empty`), non-persistent buffer save/restore, and `post_load_weights()` generically. `pipeline_cosmos3.py` and `transformer_cosmos3.py` have zero changes vs `origin/main`.

**Declarative OffloadPlan:** Models can optionally declare topology metadata via `OffloadPlan` ClassVar, replacing heuristic block discovery.

**Measured Performance (Ascend 910B3, Cosmos3-Nano 33GB):**

| Config | Per-step | Throughput | cgroup Peak | HBM/card | vs HSDP |
|--------|----------|-----------|-------------|---------|---------|
| HSDP+SP (baseline) | 870ms | 0.97 fps | — | 20.3 GB | — |
| DLO+AG DP4 (1 req) | 1,020ms | 0.81 fps | 47 GB | 12.4 GB | -17% |
| **DLO+AG DP4 (4 req)** | **1,020ms** | **3.22 fps** | **47 GB** | **12.4 GB** | **3.3×** |
| DLO no-AG DP4 | 1,877ms | 0.44 fps | 178 GB | 14.1 GB | -55% |

**Measured Performance (Ascend 910B3, Cosmos3-Super 124GB):**

| Config | cgroup Peak | Per-worker HWM | HBM/card | Video |
|--------|-------------|---------------|---------|-------|
| DLO+AG DP2 | 157 GB | 65.2 GB | ~15 GB | OK |
| DLO+AG DP4 | 172 GB | 35.5 GB | ~10 GB | OK |

**B300 (NVIDIA) Validation:**
- DLO+AG DP4, 4 concurrent T2I 1024×1024 50 steps: 4/4 succeeded, byte-identical to HSDP baseline (SHA256 match)
- 19/19 unit tests passed

---

### Sub-RFC 2: Data Parallelism for Video Generation Models

| Item | Link | Status |
|------|------|--------|
| RFC | [Issue #4707](https://github.com/vllm-project/vllm-omni/issues/4707) | Open |
| PR | [PR #4707](https://github.com/vllm-project/vllm-omni/pull/4707) | **Merged** |

**Target:** DP execution framework + bubble problem.

**Key mechanism:**
- DP multi-concurrency: scheduler batches up to `dp_size` requests per denoise step
- Dynamic load balancing: longest-processing-time-first heuristic assigns prompts to least-loaded rank
- Bubble rate drops from 20-40% to below 5%

---

### Sub-RFC 3: Iterative Activation Processing

| Item | Link | Status |
|------|------|--------|
| RFC | [Issue #5493](https://github.com/vllm-project/vllm-omni/issues/5493) | **Submitted** |
| PR | — | Not started |

**Target:** Activation memory pressure.

**Motivation:**
DP's independent replicas mean each rank processes the full sequence with all attention heads and all MoE experts. For long sequences (648K tokens), per-layer activation memory exceeds 72 GB.

**Part 1: Per-Head Iteration for Attention**

Instead of computing QKV for all heads at once, each rank iterates through heads one at a time:

```python
# Sub-RFC 3: Per-Head Iteration (Attention)
outputs = []
for head_idx in range(num_heads):  # iterate one head at a time
    Q_head = Q[:, :, head_idx, :]    # only current head's QKV in HBM
    K_head = K[:, :, head_idx, :]
    V_head = V[:, :, head_idx, :]
    out_head = attention(Q_head, K_head, V_head)  # LN + RoPE + FA per head
    outputs.append(out_head)
    # current head's intermediates freed after each iteration
return torch.cat(outputs, dim=-1)
```

**Part 2: Per-Chunk Iteration for MoE**

For MoE layers, process tokens in chunks rather than all at once:

```python
# Sub-RFC 3: Per-Chunk Iteration (MoE)
chunk_size = 4096  # tunable
outputs = []
for start in range(0, seq_len, chunk_size):
    chunk_hidden = hidden_states[start:start+chunk_size]
    expert_outputs = [expert(chunk_hidden) for expert in experts]
    chunk_output = combine(expert_outputs)
    outputs.append(chunk_output)
    # Free chunk intermediates after each iteration
return torch.cat(outputs, dim=0)
```

**Memory Impact:**

Attention (seqlen=600K, num_heads=32, head_dim=128, FP16):

| Component | Standard DP | Per-Head Iteration |
|-----------|------------|-------------------|
| QKV | 14.4 GB | 0.45 GB (1 head) |
| LN + RoPE | 14.4 GB | 0.45 GB (1 head) |
| FA intermediates | ~4.8 GB | ~0.15 GB (1 head) |
| **Peak** | **~33.6 GB** | **~1.05 GB** |

MoE (seqlen=600K, num_experts=8, hidden_dim=4096, FP16):

| Configuration | Intermediates Memory |
|--------------|---------------------|
| Standard (full seq) | 600K × 8 × 4096 × 2 = 39.3 GB |
| Chunk size = 4096 | 4096 × 8 × 4096 × 2 = 268 MB |
| Chunk size = 8192 | 536 MB |
| Chunk size = 16384 | 1.07 GB |

**Combined per-layer peak:**

| Scenario | Attention | MoE | Total |
|----------|-----------|-----|-------|
| Standard DP | 33.6 GB | 39.3 GB | ~72.9 GB |
| RFC-3 (chunk=4096) | 1.05 GB | 0.27 GB | **~1.32 GB** |

**Result:** Activation memory per layer reduced from ~73 GB to ~1.3 GB, a **~55× reduction**. No additional inter-rank communication (within-rank iteration only).

**Implementation Notes:**
- QKV projection: each head's QKV computed independently; may require custom kernels or loop-based forward
- RoPE: applied per head; computed on the fly without full Q/K allocation
- Flash Attention: works with per-head Q/K/V inputs (FlashAttention-2 on CUDA, CANN FlashAttention on NPU)
- MoE chunk size: tunable parameter balancing memory vs. compute efficiency

**Integration with Other RFCs:**
- Sub-RFC 1: weight offload frees HBM; iterative processing reduces activation memory
- Sub-RFC 2: DP execution distributes prompts; iterative processing applies within each rank
- Sub-RFC 4: independent optimization; can be combined

---

### Sub-RFC 4: Ref-Target KV Cache with Layer-Wise Caching

| Item | Link | Status |
|------|------|--------|
| RFC | [Issue #4710](https://github.com/vllm-project/vllm-omni/issues/4710) | Open |
| PR | [PR #4868](https://github.com/vllm-project/vllm-omni/pull/4868) | **Waiting for merge** |

**Target:** Redundant UND computation and KV cache pressure.

**Motivation:**
In Cosmos3's dual-pathway architecture, the reference keys/values (K_ref, V_ref) are static across denoising steps. The current pipeline recomputes cross-attention every step, wasting compute.

**Key mechanism (inspired by FullDiT2):**
- At the first denoising step (T0), compute K_ref and V_ref for every layer and cache them per layer
- In subsequent steps (T1…TN), reuse the per-layer cached K/V and bypass the reference self-attention path; only compute the noisy-to-all attention with the evolving generation queries

**Expected outcome:** Per-step cross-attention compute for the Ref-related part reduced by ~50% (for a typical 20-step diffusion). Memory overhead ~3.2 GB for an 80-layer 64B model.

## Dependency Graph

```
                    ┌─────────────────┐
                    │   Sub-RFC 1     │  ← PR #5397 (waiting for merge)
                    │ (Weight Offload)│
                    └────────┬────────┘
                             │
                             ▼
                    ┌─────────────────┐
                    │   Sub-RFC 2     │  ← PR #4707 (merged)
                    │   (Async DP)    │
                    └────────┬────────┘
                             │
              ┌──────────────┴──────────────┐
              │                             │
              ▼                             ▼
    ┌─────────────────┐           ┌─────────────────┐
    │   Sub-RFC 3     │           │   Sub-RFC 4     │  ← PR #4868 (waiting)
    │(Activation Proc)│           │(Ref-Target KV)  │
    │ ← Issue #5493   │           │                 │
    └─────────────────┘           └─────────────────┘
```

## Combined Performance Impact (64B, 8× GPUs)

| Metric | SP | DP (Static) | DP (All RFCs) |
|--------|----|-----------|--------------|
| Communication | O(seq_len) | Zero | Zero |
| GEMM efficiency | Low (~20%) | High (~50%) | High (~60%) |
| Bubble rate | 0% | 20-40% | <5% |
| Device HBM per rank | ~129 GB | 129 GB (OOM) | ~10 GB |
| Host memory (8 ranks) | N/A | 1,032 GB (OOM) | ~129 GB |
| Activation per layer | ~72 GB | ~72 GB | ~1.3 GB |
| Cross-attention compute (per step) | baseline | baseline | ~50% reduction (Ref part) |

## Hardware Support

| Platform | Sub-RFC 1 | Sub-RFC 2 | Sub-RFC 3 | Sub-RFC 4 |
|----------|-----------|-----------|-----------|-----------|
| NVIDIA GPU | ✅ | ✅ | ✅ | ✅ |
| Ascend NPU | ✅ | ✅ | ✅ | ✅ |

## Recipes

| Device | DP Size | Command |
|--------|---------|---------|
| 2× H200 | 2 | `--enable-distributed-layerwise-offload --data-parallel-size 2` |
| 4× H100 | 4 | `--enable-distributed-layerwise-offload --data-parallel-size 4` |
| 8× A100 | 8 | `--enable-distributed-layerwise-offload --data-parallel-size 8` |
| 4× H100 + Ref-Target KV | 4 | `--enable-distributed-layerwise-offload --data-parallel-size 4 --enable-ref-target-kv` |
| 2× H200 + Ref-Target KV (CPU offload) | 2 | `--enable-distributed-layerwise-offload --data-parallel-size 2 --enable-ref-target-kv --ref-cache-on-cpu` |

### Proposed Change.

Please provide the detailed design document of the RFC using the [template](https://docs.google.com/document/d/1jcgR3cDaUQH3VczD4ZcKaJAoYWHjCmnYzHkCNyz-9fk/edit?usp=sharing).

### CC List.

@hsliuustc0106 @Gaohan123 @david6666666 @bjf-frz 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
